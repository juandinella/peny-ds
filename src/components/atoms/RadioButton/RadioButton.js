import React from 'react'
import PropTypes from 'prop-types'
import styles from './RadioButton.styles'
import Skeleton from '../Skeleton'
import Utils from '../../../utils'

/**
 * Provides single user input from a selection of options.
 */

const RadioButton = ({
  checked,
  disabled,
  id,
  radioName,
  label,
  value,
  onChange,
  shouldRender,
}) => {
  return (
    <Skeleton
      id={`${id}_skeleton`}
      show={!Utils.emptyValue(shouldRender) && !shouldRender()}
    >
      <div css={styles} data-testid={id}>
        <input
          type="radio"
          name={radioName}
          id={id}
          onChange={disabled ? null : onChange}
          defaultChecked={checked}
          disabled={disabled}
          value={value}
        />
        <label htmlFor={id}>{label}</label>
      </div>
    </Skeleton>
  )
}

RadioButton.displayName = 'RadioButton'

RadioButton.propTypes = {
  /** RadioButton's id */
  id: PropTypes.string,
  /** RadioButton's value */
  value: PropTypes.string,
  /** This Boolean attribute to display a checked radio button */
  checked: PropTypes.bool,
  /** RadioButton's name */
  radioName: PropTypes.string.isRequired,
  /** RadioButton's label */
  label: PropTypes.string.isRequired,
  /** Callback for RadioButton */
  onChange: PropTypes.func,
  /** Disabled RadioButton */
  disabled: PropTypes.bool,
  /**
   * Function that decides if the component should be rendered.
   *
   * @return {boolean} false if component should not be rendered and show skeleton instead. Component renders otherwise.
   */
  shouldRender: PropTypes.func,
}

RadioButton.defaultProps = {
  checked: false,
}

export default RadioButton
