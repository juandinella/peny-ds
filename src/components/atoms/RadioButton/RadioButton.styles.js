import { css } from 'styled-components'

export default css`
  display: flex;
  align-items: center;
  cursor: pointer;

  input[type='radio'] {
    -webkit-appearance: none;
    -moz-appearance: none;
    appearance: none;
    display: inline-block;
    width: var(--spacing-sm);
    height: var(--spacing-sm);
    min-width: var(--spacing-sm);
    min-height: var(--spacing-sm);
    padding: var(--spacing-xxxs);
    background-clip: content-box;
    border: 1px solid var(--color-neutral-400);
    border-radius: 50%;
    cursor: pointer;
    position: static;
    transition: background-color 0.15s ease, border-color 0.15s ease;
  }

  input[type='radio']:checked {
    background-color: var(--color-primary-main);
    border-color: var(--color-primary-main);
  }

  label {
    margin-left: var(--spacing-xxs);
    cursor: pointer;
    width: 100%;
    display: inline-flex;
    align-items: center;
  }
`
