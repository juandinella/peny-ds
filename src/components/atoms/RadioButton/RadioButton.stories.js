import RadioButton from './RadioButton'
import { getTemplate } from '../../../helpers/storybook'

const Template = getTemplate(RadioButton)

export default {
  title: 'Atoms/RadioButton',
  component: RadioButton,
  args: {
    label: 'Im a Radio Button',
    id: 'radio-id',
    radioName: 'radio-name',
    onChange: () => true,
  },
  argTypes: {},
}

export const Default = Template.bind({})

export const Skeleton = Template.bind({})
Skeleton.args = { shouldRender: () => false }
