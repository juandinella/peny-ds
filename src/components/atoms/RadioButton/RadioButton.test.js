import React from 'react'

import { render, fireEvent, cleanup } from '@testing-library/react'

import RadioButton from './RadioButton'

import initStoryshots from '@storybook/addon-storyshots'

initStoryshots({
  storyKindRegex: /^Atoms\/RadioButton$/,
})

const componentProps = {
  id: 'test-id',
  radioName: 'test-name',
  label: 'test-label',
  checked: false,
  disabled: false,
  onChange: jest.fn(),
}

afterEach(cleanup)

describe('<RadioButton />', () => {
  test('should render component correctly', () => {
    const { queryByTestId } = render(<RadioButton {...componentProps} />)

    const radioButton = queryByTestId(componentProps.id)

    expect(radioButton).toBeTruthy()
  })

  test('should render skeleton', () => {
    const { queryByTestId } = render(
      <RadioButton {...componentProps} shouldRender={() => false} />
    )

    expect(queryByTestId(`${componentProps.id}_skeleton`)).toBeTruthy()
  })

  test('input should have radio name', () => {
    const { container } = render(<RadioButton {...componentProps} />)

    const input = container.querySelector('input')

    expect(input.name).toEqual(componentProps.radioName)
  })

  test('should render content correctly', () => {
    const { container } = render(<RadioButton {...componentProps} />)

    const label = container.querySelector('label')

    expect(label.textContent).toEqual(componentProps.label)
  })

  test('should call function on click', () => {
    const { container } = render(<RadioButton {...componentProps} />)

    fireEvent.click(container.getElementsByTagName('input')[0])

    expect(componentProps.onChange).toHaveBeenCalledTimes(1)
  })
})
