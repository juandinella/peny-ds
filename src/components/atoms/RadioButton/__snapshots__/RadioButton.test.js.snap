// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storyshots Atoms/RadioButton Default 1`] = `
<div
  style={
    Object {
      "alignContent": "flex-start",
      "display": "flex",
      "flexDirection": "column",
      "flexWrap": "wrap",
      "gap": "10px 30px",
      "height": "100%",
      "justifyContent": "flex-start",
      "maxHeight": "auto",
    }
  }
>
  <div
    className="sc-6kv2s7-0 jTNQcl"
    data-testid="radio-id"
  >
    <input
      defaultChecked={false}
      id="radio-id"
      name="radio-name"
      onChange={[Function]}
      type="radio"
    />
    <label
      htmlFor="radio-id"
    >
      Im a Radio Button
    </label>
  </div>
</div>
`;

exports[`Storyshots Atoms/RadioButton Skeleton 1`] = `
<div
  style={
    Object {
      "alignContent": "flex-start",
      "display": "flex",
      "flexDirection": "column",
      "flexWrap": "wrap",
      "gap": "10px 30px",
      "height": "100%",
      "justifyContent": "flex-start",
      "maxHeight": "auto",
    }
  }
>
  <div
    className="sc-kjnynt-0 jfZhXD"
    data-testid="radio-id_skeleton"
    id="radio-id_skeleton"
    style={
      Object {
        "height": undefined,
        "width": undefined,
      }
    }
  >
    <div
      className="sc-6kv2s7-0 jTNQcl"
      data-testid="radio-id"
    >
      <input
        defaultChecked={false}
        id="radio-id"
        name="radio-name"
        onChange={[Function]}
        type="radio"
      />
      <label
        htmlFor="radio-id"
      >
        Im a Radio Button
      </label>
    </div>
  </div>
</div>
`;
